File: sudoku/sudoku.c
/*
 * sudoku - Solve and create sudoku boards! For more see
 * DESIGN.md, REQUIREMENTS.md, IMPLEMENTATION.md, and
 * README.md
 *
 * usage: sudoku [mode]
 * 
 * mode can be "solver" or "create" or "play"
 *
 * CS50, Spring 2022
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <time.h>
#include <unistd.h> // read, write, close
#include <string.h> // memcpy, memset
#include <netdb.h>  // socket-related structures
#include "../libcs50/tile.h"
#include "../libcs50/grid.h"
#include "../libcs50/memory.h"
#include "../libcs50/file.h"
#include "../libcs50/numbers.h"

/**************** file-local constants ****************/
#define BUFSIZE 1024 // read/write buffer size
#define HOSTNAME "plank"
#define PORT 3345

// function prototypes 
void play_grid(int N, grid_t *grid_answer, grid_t *grid_puzzle, grid_t *grid_user);
grid_t *get_grid();
int get_sudoku_size();

    int main(int argc, char *argv[])
{
    // check number of arguments
    if (argc != 2) {
        fprintf(stderr, "Usage: ./sudoku mode\n");
        exit(1);
    }

    if (!strcmp(argv[1], "create")) {
        printf("Sudoku mode: create\n");
        int N = 0;
        grid_t *grid = get_grid(&N);

        grid_print(grid, false);

        printf("Empty tiles count: %d\n", grid_empty_cnt(grid));
        grid_delete(grid);

    } else if (!strcmp(argv[1], "solver")) {
        printf("Sudoku mode: solver\n");

        int N = get_sudoku_size();
        grid_t *grid = grid_new(N);
        // grid_print(grid, false);

        // load sudoku from input
        if (! grid_load(grid)) {
            exit(4);
        }
        grid_fprint(grid, false);

        int sol = grid_solve(grid);
        if (sol == 0) {
            printf("Sudoku solution not found\n");
        } else if (sol == 1){
            printf("Unique Sudoku Solution found:\n");
            grid_print(grid, false);
            grid_fprint(grid, true);
        }
        else
        {
            printf("Multiple Sudoku Solutions found:\n");
            grid_print(grid, false);
            grid_fprint(grid, true);
        }
        grid_delete(grid);

    } else if (!strcmp(argv[1], "play")) {
        // generate a sudoku puzzle
        int N = 0;
        grid_t *grid_puzzle = get_grid(&N);
        grid_t *grid_answer = grid_new(N);
        grid_copy(grid_puzzle, grid_answer);

        grid_solve(grid_answer);

        grid_print(grid_puzzle, true);

        grid_t *grid_user = grid_new(N); // stores the user input
        assertp(grid_user, "Failed to allocate memory for grid");
        grid_copy(grid_puzzle, grid_user);

        play_grid(N, grid_answer, grid_puzzle, grid_user);

        grid_delete(grid_answer);
        grid_delete(grid_puzzle);
        grid_delete(grid_user);
    } else {
        fprintf(stderr, "Unknown mode. Mode can be solver or create or play\n");
        exit(5);
    }

    return 0;
}

// read user input and play the grid
void play_grid(int N, grid_t *grid_answer, grid_t *grid_puzzle, grid_t *grid_user)
{
    int empty_cnt = grid_empty_cnt(grid_puzzle);
    int wrong_cnt = empty_cnt;
    printf("Empty tiles count: %d\n", empty_cnt);

    while (wrong_cnt > 0) { // while the user has not found the correct answer
        while (empty_cnt > 0) { // while the user sudoku has not been filled
            int i;
            int j;
            int num;
            char excess;

            // read and validate input
            printf("Enter your next step (row col num):");
            char *line = freadlinep(stdin);
            if (!strcmp(line, "q")) {
                printf("Quit program\n");
                free(line);
                return;
            }
            if (sscanf(line, "%d %d %d%c", &i, &j, &num, &excess) != 3) {
                fprintf(stderr, "Invalid input. Try again\n");
                free(line);
                continue;
            }
            if ((i < 0) || (i >= N)) {
                fprintf(stderr, "Row %d must be between [0, %d]. Try again\n", i, N-1);
                free(line);
                continue;
            }
            if ((j < 0) || (j >= N)) {
                fprintf(stderr, "Col %d must be between [0, %d]. Try again\n", j,  N-1);
                free(line);
                continue;
            }
            if ((num < 1) || (num > N)) {
                fprintf(stderr, "Num %d must be between [1, %d]. Try again\n", num, N);
                free(line);
                continue;
            }
            if (grid_get_status(grid_user, i, j)) {
                fprintf(stderr, "Tile (%d, %d) is fixed (a clue). Try again\n", i, j);
                free(line);
                continue;
            }
            // store the previous number in the specified tile
            int prev_num = grid_get_value(grid_user, i, j);
            // set the value in the user grid
            grid_set_tile(grid_user, i, j, false, num);

            // increment empty_cnt if a nonzero tile is set to zero
            if  ((prev_num != 0) && (num == 0)) {
                empty_cnt++;
            // decrement empty_cnt if a zero tile is set to nonzero
            } else if  ((prev_num == 0) && (num != 0)) {
                empty_cnt--;
            } // otherwise no changes to empty_cnt

            int correct_num = grid_get_value(grid_answer, i, j);
            // decrement wrong_cnt if an incorrect tile is set to correct
            if ((prev_num != correct_num) && (num == correct_num)) {
                grid_set_tile(grid_puzzle, i, j, false, num);
                wrong_cnt--;
            // increment wrong_cnt if a correct tile is set to incorrect
            } else if ((prev_num == correct_num) && (num != correct_num)) {
                grid_set_tile(grid_puzzle, i, j, false, 0);
                wrong_cnt++;
            } // otherwise no changes to wrong_cnt

            free(line);
            grid_print(grid_user, true);
        }
        if (wrong_cnt == 0) {
            printf("Success! You solved the sudoku!\n");
            break;
        } else {
            printf("Uh oh...Incorrect solution. We've kept the correct tiles. Try again!\n");
            
            // restart game
            // copy puzzle grid (that has correct user input) to the user grid
            grid_copy(grid_puzzle, grid_user);
            grid_print(grid_user, true);
            empty_cnt = grid_empty_cnt(grid_puzzle);
            wrong_cnt = empty_cnt;
            printf("Empty tiles count: %d\n", empty_cnt);
        }
    }
}


grid_t *get_grid(int *N){
    // code taken from inserver/inclient assignment
    *N = get_sudoku_size();
    char *hostname = HOSTNAME; // server hostname
    int port = PORT;           // server port

    // 1. Create socket
    int comm_sock = socket(AF_INET, SOCK_STREAM, 0);
    if (comm_sock < 0)
    {
        perror("opening socket");
        exit(2);
    }

    // 2. Initialize the fields of the server address
    struct sockaddr_in server; // address of the server
    server.sin_family = AF_INET;
    server.sin_port = htons(port);
    // Look up the hostname specified on command line
    struct hostent *hostp = gethostbyname(hostname); // server hostname
    if (hostp == NULL)
    {
        fprintf(stderr, "unknown host '%s'\n", hostname);
        exit(3);
    }
    memcpy(&server.sin_addr, hostp->h_addr_list[0], hostp->h_length);

    // 3. Connect the socket to that server
    if (connect(comm_sock, (struct sockaddr *)&server, sizeof(server)) < 0)
    {
        perror("connecting stream socket");
        exit(4);
    }
    printf("Connected!\n");

    char buf[BUFSIZE];       // a buffer for reading data from stdin
    memset(buf, 0, BUFSIZE); // clear up the buffer

    freopen("/dev/null", "a", stdout);
    setbuf(stdout, buf);
    printf("%d", *N);
    freopen("/dev/tty", "a", stdout);


    write(comm_sock, buf, BUFSIZE-1);

    memset(buf, 0, BUFSIZE); // clear up the buffer

    read(comm_sock, buf, BUFSIZE - 1);

    grid_t *grid = grid_new(*N);
    assertp(grid, "Failed to allocate memory for grid");

    grid_load_string(grid, buf);
    
    close(comm_sock);

    return grid;
}

// get and check the size of the sudoku from user input
int get_sudoku_size()
{

    printf("Enter the size of the sudoku (4, 9, 16, ...): ");
    int N;
    char excess;
    char *line = freadlinep(stdin);

    if (sscanf(line, "%d%c", &N, &excess) != 1)
    {
        fprintf(stderr, "Invalid input: must be an integer\n");
        free(line);
        exit(3);
        return 0;
    }

    if ((N <= 0) || (!is_perfect_square(N)))
    {
        fprintf(stderr, "Invalid input: must be a positive perfect square\n");
        free(line);
        exit(4);
        return 0;
    }
    printf("Sudoku size: %d\n", N);
    free(line);
    return N;
}

File: main/sudoku.py
"""
The sudoku program using the command line as user interface
Usage: python sudoku.py [mode]
[mode] can be create, solver or play

Guhui Zhang, 2022
"""

import sys
sys.path.insert(0, '../lib')
import random
import number
from grid import *


def get_sudoku_size():
    """
    reads the size of the sudoku from stdin,
    checks whether the number is a positive perfect square integer

    :return: the inputted size of the sudoku; -1 if the number is invalid
    """
    try:
        n = int(input('Enter the size of the sudoku (4, 9, 16, ...): '))
    except Exception as e:
        print('Invalid input: must be an integer', file=sys.stderr)
        print(e, file=sys.stderr)
        return -1

    if (n <= 0) or (not number.is_perfect_square(n)):
        print('Invalid input: must be a positive perfect square', file=sys.stderr)
        return -1

    print(f'Sudoku size: {n}')
    return n


def play_grid(n, grid_answer, grid_puzzle, grid_user):
    """
    implements the play mode

    :param n: the size of the sudku
    :param grid_answer: the grid containing the answer to the sudoku puzzle
    :param grid_puzzle: the grid containing the sudoku puzzle
    :param grid_user: the grid containing the user input to the puzzle
    :return: none
    """
    empty_cnt = grid_puzzle.grid_empty_cnt()
    wrong_cnt = empty_cnt
    print(f'Empty tiles count: {empty_cnt}')

    # stores the clues; intended for giving a clue to the user
    clues = [(i, j) for i in range(n) for j in range(n) if not grid_puzzle.grid_get_status(i, j)]

    while wrong_cnt > 0:  # while the puzzle is not correctly filled
        while empty_cnt > 0:  # while the puzzle is not filled
            line = input("Enter your next step (row col num), or type 'clue' or 'q':")
            if line == 'q':
                print('Quit program')
                return

            if line == 'clue':  # gives a random clue and fill it in the user grid
                temp = random.randint(0, len(clues)-1)
                i, j = clues.pop(temp)
                num = grid_answer.grid_get_value(i, j)
                print(f'Clue: {i} {j} {num}')
                grid_user.grid_set_tile(i, j, False, num)
            else:
                try:  # the input should be row col num
                    args = [int(arg) for arg in line.split()]
                    i = args[0]
                    j = args[1]
                    num = args[2]
                except Exception as e:
                    print('Invalid input. Try again', file=sys.stderr)
                    print(e, file=sys.stderr)
                    return

            if (i < 0) or (i >= n):
                print(f'Row {i} must be between [0, {n-1}]. Try again', file=sys.stderr)
                continue

            if (j < 0) or (j >= n):
                print(f'Column {j} must be between [0, {n-1}]. Try again', file=sys.stderr)
                continue

            if (num < 1) or (num > n):
                print(f'Number {num} must be between [1, {n}]. Try again', file=sys.stderr)
                continue

            if grid_user.grid_get_status(i, j):
                print(f'Tile ({i}, {j}) is a fixed clue. Try again', file=sys.stderr)

            # store the previous number in the specified tile
            prev_num = grid_user.grid_get_status(i, j)
            # set the value in the user grid
            grid_user.grid_set_tile(i, j, False, num)

            # increment empty_cnt if a nonzero tile is set to zero
            if (prev_num != 0) and (num == 0):
                empty_cnt += 1
            # decrement empty_cnt if a zero tile is set to nonzero
            elif (prev_num == 0) and (num != 0):
                empty_cnt -= 1

            correct_num = grid_answer.grid_get_value(i, j)
            # decrement wrong_cnt if an incorrect tile is set to correct
            if (prev_num != correct_num) and (num == correct_num):
                grid_puzzle.grid_set_tile(i, j, False, num)
                wrong_cnt -= 1
            # increment wrong_cnt if a correct tile is set to incorrect
            elif (prev_num == correct_num) and (num != correct_num):
                grid_puzzle(i, j, False, 0)
                wrong_cnt += 1

            grid_user.grid_print(play_mode=True)

        if wrong_cnt == 0:
            print('Success! You solved the sudoku!')
        else:
            print("Uh oh...Incorrect solution. We've kept the correct tiles. Try again!")

            # restart game
            # copy puzzle grid (that has correct user input) to the user grid
            grid_puzzle.grid_copy(grid_user)
            grid_user.grid_print(play_mode=True)
            empty_cnt = grid_puzzle.grid_empty_cnt()
            wrong_cnt = empty_cnt
            print(f'Empty tiles count: {empty_cnt}')


argc = len(sys.argv)
if argc != 2:  # check number of arguments
    print('Usage: python sudoku.py mode', file=sys.stderr)
    exit(1)

if sys.argv[1] == 'create':
    print('Sudoku mode: create')
    n = get_sudoku_size()
    if n == -1:
        exit(2)

    grid = Grid(n)
    grid.grid_solve()
    grid.grid_random_clue()
    grid.grid_print()

elif sys.argv[1] == 'solver':
    print('Sudoku mode: solver')
    n = get_sudoku_size()
    if n == -1:
        exit(2)

    grid = Grid(n)
    if not grid.grid_load():
        exit(3)

    sol = grid.grid_solve()
    if sol == 0:
        print('Sudoku solution not found')
    else:
        print(f'{sol} solution(s) found:')
        grid.grid_print()

elif sys.argv[1] == 'play':
    n = get_sudoku_size()
    grid_puzzle = Grid(n)  # stores the puzzle
    grid_answer = Grid(n)  # stores the answer
    grid_user = Grid(n)  # stores the user input

    grid_puzzle.grid_solve()
    grid_puzzle.grid_copy(grid_answer)

    grid_puzzle.grid_random_clue()
    grid_puzzle.grid_print(play_mode=True)
    grid_puzzle.grid_copy(grid_user)

    play_grid(n, grid_answer, grid_puzzle, grid_user)

else:
    print('Unknown mode. Mode can be solver or create or play', file=sys.stderr)
    exit(4)
